### PR TITLE
[CALCITE-3935] Enhance Join-Materialization, support to pull-up filters under join of left or right (xurenhe)

### PR DESCRIPTION
**Pull request by wojustme**
Fix Bug as Follows
https://issues.apache.org/jira/browse/CALCITE-3935

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 1]
```diff
@@ -1202,8 +1202,9 @@ private JoinOnLeftCalcToJoinUnifyRule() {
       if (joinRelType == null) {
         return null;
       }
-      if (joinRelType != JoinRelType.INNER
-          && !(joinRelType.isOuterJoin() && qInput0Cond.isAlwaysTrue())) {
+      // If JoinType is inner-join or out-join,
+      // Calc which is bottom join can be pulled up.
+      if (joinRelType != JoinRelType.INNER && !joinRelType.isOuterJoin()) {
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:1207]: What if it's a right Join but with Calc as the left child ?
If above concern is valid, please add a corresponding test.

**Comment by jinxing64** [comment]
Please refine the commit message.

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 2]
```diff
@@ -2011,6 +2006,38 @@ public static boolean equalType(String desc0, MutableRel rel0, String desc1,
...
+  /**
+   * Check filter bottom join can be pull-up,
+   * when meeting JoinOnCalc of query unify to Join of target.
+   * Such as {@link JoinOnLeftCalcToJoinUnifyRule} <br/>
+   * {@link JoinOnRightCalcToJoinUnifyRule} <br/>
+   * {@link JoinOnCalcsToJoinUnifyRule} <br/>
+   */
+  private static boolean canPullUpFilterBottomJoin(JoinRelType joinType,
+                                                   RexNode leftFilterRexNode,
+                                                   RexNode rightFilterRexNode) {
```
Comment (yanlin-Lynn) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:2018]: we do not format like this

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 3]
```diff
@@ -2011,6 +2006,38 @@ public static boolean equalType(String desc0, MutableRel rel0, String desc1,
     return RelOptUtil.equal(desc0, rel0.rowType, desc1, rel1.rowType, litmus);
   }
 
+  /**
+   * Check filter bottom join can be pull-up,
```
Comment (yanlin-Lynn) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:2010]: Check if filter under join can be pulled up

Comment (wojustme) [reply core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:2078]: ok, thanks for your remind.

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java (hunk 2 above)
Comment (wojustme) [reply core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:2086]: ok, I format it.

**Comment by wojustme** [comment]
@xy2953396112 
Hi, Could take your time, review this pr?

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 4]
```diff
@@ -1388,16 +1388,11 @@ private JoinOnCalcsToJoinUnifyRule() {
       final RexBuilder rexBuilder = call.getCluster().getRexBuilder();
 
       // Try pulling up MutableCalc only when:
```
Comment (yanlin-Lynn) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:1390]: You should refine this note.

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 5]
```diff
@@ -1303,8 +1303,8 @@ private JoinOnRightCalcToJoinUnifyRule() {
       if (joinRelType == null) {
         return null;
       }
-      if (joinRelType != JoinRelType.INNER
-          && !(joinRelType.isOuterJoin() && qInput1Cond.isAlwaysTrue())) {
+      // Check filter bottom join can be pulled up.
```
Comment (yanlin-Lynn) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:1306]: ``` Check if filter under join can be pulled up```, keep the same with the note for this method.

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 6]
```diff
@@ -2083,6 +2078,38 @@ public static boolean equalType(String desc0, MutableRel rel0, String desc1,
...
+   * {@link JoinOnCalcsToJoinUnifyRule} <br/>
+   */
+  private static boolean canPullUpFilterUnderJoin(JoinRelType joinType,
+                                                  @Nullable RexNode leftFilterRexNode,
+                                                  @Nullable RexNode rightFilterRexNode) {
+    if (joinType == JoinRelType.INNER) {
+      return true;
+    }
+    if (joinType == JoinRelType.LEFT) {
+      if (rightFilterRexNode == null || rightFilterRexNode.isAlwaysTrue()) {
```
Comment (yanlin-Lynn) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:2095]: Suggest you write like this.
```
if (joinType == JoinRelType.LEFT) && (rightFilterRexNode == null || rightFilterRexNode.isAlwaysTrue()) {
  return true;
}
```

**Comment by wojustme** [comment]
@yanlin-Lynn 
Thanks for your advise.
Please review again.

File: core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java [hunk 7]
```diff
@@ -2083,6 +2075,35 @@ public static boolean equalType(String desc0, MutableRel rel0, String desc1,
...
+  /**
+   * Check if filter under join can be pulled up,
+   * when meeting JoinOnCalc of query unify to Join of target.
+   * Such as {@link JoinOnLeftCalcToJoinUnifyRule} <br/>
+   * {@link JoinOnRightCalcToJoinUnifyRule} <br/>
+   * {@link JoinOnCalcsToJoinUnifyRule} <br/>
+   */
+  private static boolean canPullUpFilterUnderJoin(JoinRelType joinType,
+                                                  @Nullable RexNode leftFilterRexNode,
+                                                  @Nullable RexNode rightFilterRexNode) {
```
Comment (yanlin-Lynn) [inline core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java:2087]: fix this indentation please.

Comment (wojustme) [reply core/src/main/java/org/apache/calcite/plan/SubstitutionVisitor.java]: @yanlin-Lynn 
Thanks for reminding.
Please review again.

**Comment by yanlin-Lynn** [comment]
LGTM, trigger the checks again please.

**Pull request merged**